dvc/tree/dvc.py

```diff
@@ -45,18 +47,39 @@ def _is_cached(out):
 
         return outs
 
-    def _get_granular_checksum(self, path, out, remote=None):
+    def _get_granular_hash(self, path, out, remote=None):
         assert isinstance(path, PathInfo)
+
         if not self.fetch and not self.stream:
             raise FileNotFoundError
+
+        # NOTE: use string paths here for performance reasons
+        self._update_dir_entry_hashes(out, remote)
+        path_str = str(path)
+        hash_info = self._dir_entry_hashes.get(path_str)
+        if hash_info:
+            return hash_info
+        raise FileNotFoundError
+
+    def _update_dir_entry_hashes(self, out, remote):
+        # cache the most recently used output dir cache to avoid expensive
+        # repeated lookups of individual files within the same large output dir
         dir_cache = out.get_dir_cache(remote=remote)
```

Comment:
So this particular one is slow, right?

Note that we have dir_cache caching in the Cache itself https://github.com/iterative/dvc/blob/7b590e7c5d9361a611443c057313a6686b146f50/dvc/cache/base.py#L64 , wonder why that one wasn't enough and we need to introduce another level.

Reply:
Probably because iterating over the whole dir_cache to find one hash is slow. Do I understand correctly that `walk` was slow? If so, would it be possible to bulk up the operation somehow so we don't `metadata()` for each file in a giant dataset?

Reply:
Right, `walk` was slow because we would re-iterate over the whole dir_cache when looking for each individual entry/file hash